pygpu/collectives.pyx

```diff
@@ -0,0 +1,299 @@
+            comm_generate_id(self.context.ctx, self)
+            res = self.comm_id.internal[:GA_COMM_ID_BYTES]
+            return res
+
+        def __set__(self, unsigned char[:] bytearr):
+            cdef int length
+            length = bytearr.shape[0]
+            if length < GA_COMM_ID_BYTES:
+                raise ValueError, "GpuComm clique id must have length %d bytes" % (GA_COMM_ID_BYTES)
+            memcpy(self.comm_id.internal, <char*>&bytearr[0], GA_COMM_ID_BYTES)
+
+
+cdef class GpuComm:
+    """
+    """
+    def __dealloc__(self):
+        gpucomm_free(self.c)
+
+    def __reduce__(self):
+        raise RuntimeError, "Cannot pickle GpuComm object"
+
+    def __cinit__(self, GpuCommCliqueId cid not None, int ndev, int rank):
+        self.context = cid.context
+        cdef int err
+        err = gpucomm_new(&self.c, self.context.ctx, self.cid.comm_id,
+                          ndev, rank)
+        if err != GA_NO_ERROR:
+            raise get_exc(err), gpucontext_error(self.context.ctx, err)
+
+    def get_count(self):
+        cdef int gpucount
+        comm_get_count(self, &gpucount)
+        return gpucount
+
+    def get_rank(self):
+        cdef int gpurank
+        comm_get_rank(self, &gpurank)
+        return gpurank
+
+    def reduce(self, GpuArray src not None, op, GpuArray dest=None, int root=-1):
+        if dest is None:
+            if root != -1:
+                return comm_reduce_from(self, src, to_reduce_opcode(op), root)
+            else:
+                return pygpu_make_reduced(self, src, to_reduce_opcode(op))
+        if root == -1:
+            comm_get_rank(self, &root)
+        return comm_reduce(self, src, dest, to_reduce_opcode(op), root)
+
+    def all_reduce(self, GpuArray src not None, op, GpuArray dest=None):
+        if dest is None:
+            return pygpu_make_all_reduced(self, src, to_reduce_opcode(op))
+        return comm_all_reduce(self, src, dest, to_reduce_opcode(op))
+
+    def reduce_scatter(self, GpuArray src not None, op, GpuArray dest=None):
+        if dest is None:
+            return pygpu_make_reduce_scattered(self, src, to_reduce_opcode(op))
+        return comm_reduce_scatter(self, src, dest, to_reduce_opcode(op))
+
+    def broadcast(self, GpuArray array not None, int root=-1):
+        if root == -1:
+            comm_get_rank(self, &root)
+        return comm_broadcast(self, array, root)
+
+    def all_gather(self, GpuArray src not None, GpuArray dest=None,
+                   unsigned int nd_up=1):
+        if dest is None:
+            return pygpu_make_all_gathered(self, src, nd_up)
+        return comm_all_gather(self, src, dest)
+
+
+cdef dict TO_RED_OP = {
+    '+': GA_SUM,
+    "sum": GA_SUM,
+    "add": GA_SUM,
+    '*': GA_PROD,
+    "prod": GA_PROD,
+    "product": GA_PROD,
+    "max": GA_MAX,
+    "maximum": GA_MAX,
+    "min": GA_MIN,
+    "minimum": GA_MIN,
+    }
+
+cdef int to_reduce_opcode(op) except -1:
+    if isinstance(op, int):
+        return op
```

Comment:
The integer values are never exposed anywhere so if the user uses some they are probably wrong.

Either expose the `GA_...`  values or remove this part.